Factor mechanics code from IBFEMethod into FEMechanicsBase

boyceg opened the pull request:
<!--
This template should be included in all pull requests. Items in the list should
either be completed by the original author or explicitly dismissed by one of the
IBAMR principal developers.

IBAMR is a community effort and it wouldn't exist without people contributing
code. Thanks in advance for helping to make IBAMR better!
-->

### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format run?
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [x] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

boyceg commented:
@drwells, I think this is passing most of the tests that usually work for me on MacOS. The restart files have changed because of the refactoring, and so I think it will be necessary to update those tests.

boyceg commented:
Ah, the problem with restart is that I screwed up `FEMechanicsBase::putToDatabase()`. Should be fixed now...

drwells commented on include/ibamr/IBFEMethod.h:
```diff
@@ -302,280 +302,72 @@ class IBFEDirectForcingKinematics;
      */
     IBFEMethod(const std::string& object_name,
-               SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db,
+               const SAMRAI::tbox::Pointer<SAMRAI::tbox::Database>& input_db,
```
:tada:

drwells commented on include/ibamr/FEMechanicsBase.h:
```diff
@@ -0,0 +1,606 @@
+    std::vector<std::shared_ptr<IBTK::FEProjector> > d_fe_projectors;
+
+    /// Number of parts owned by the present object.
+    const unsigned int d_num_parts = 1;
```
This didn't make any sense in IBFEMethod either so we should remove it.

drwells commented on include/ibamr/IBFEMethod.h:
```diff
@@ -636,8 +428,8 @@ class IBFEMethod : public IBStrategy
-    virtual void inactivateLagrangianStructure(int structure_number = 0,
-                                               int level_number = std::numeric_limits<int>::max()) override;
+    void inactivateLagrangianStructure(int structure_number = 0,
+                                       int level_number = std::numeric_limits<int>::max()) override;
```
Can we move this and related functions into `FEMechanicsBase`? Since that class already stores data structures per part, it makes more sense that these should be there.

boyceg replied:
I can move the data structure, but `inactivateLagrangianStructure` and related functions are in `IBStrategy`, so we cannot completely get rid of them in `IBFEMethod`.

drwells replied:
Oh duh. Lets definitely not move this into `FEMechanicsBase` - its too complicated to think about which parent class has the right virtual function definition.

drwells commented on src/IB/IBFEMethod.cpp:
```diff
@@ -885,8 +693,7 @@ IBFEMethod::backwardEulerStep(const double current_time, const double new_time)
     }
-    return;
-} // backwardEulerStep
+}
```
Could we please avoid these types of code updates in this PR? I like this change but its making this patch bigger than it needs to be.

boyceg replied:
I knew you would say this, but let me tell you why I went ahead and did it: I am now running `clang-tidy` interactively, and getting rid of these `return` statements makes `clang-tidy` happy. (And I figured if I am wiping out the unneeded `return` I also might as well get rid of the unneeded comment, which dates back to days when I did not have good block matching tools in my editor.)

drwells commented on include/ibamr/IBFEMethod.h:
```diff
@@ -1162,47 +866,45 @@ class IBFEMethod : public IBStrategy
+     * Convenience pointer for object managing access to libMesh system vectors
+     * for the force.
      */
-    std::vector<libMesh::PetscVector<double>*> d_Q_half_vecs, d_Q_rhs_vecs, d_Q_IB_ghost_vecs;
```
I deliberately left all the fluid source/sink vectors alone since we don't have any tests - are you sure that we should proceed with this change at this time? I would prefer to handle this later.

boyceg replied:
Yes let's do it now.

boyceg commented:
Hmmm, I thought that tests were working. I will take another look later — but could be a while — I need to bake a birthday cake and get started on a grant tonight. :-D

boyceg commented:
Well, I got the cake baked. I think I also fixed most of the problems that were causing tests to fail, and reverted some of the "improvements" that were getting in the way of review.

@drwells can you take another look?

drwells commented:
I'm on it!

drwells commented:
The tests now pass, so that's a good start. I'm reading more carefully now.

drwells commented on src/IB/IBFEMethod.cpp:
```diff
@@ -354,24 +343,50 @@ build_ib_ghosted_system_data(std::vector<SystemData>& ghosted_system_data,
+        auto& ghosted_vector = dynamic_cast<PetscVector<Number>&>(system.get_vector(vector_name));
+        insert_parallel_into_ghosted(parallel_vector, ghosted_vector);
+    }
+};
```
AFAICT we shouldn't need to use these in IBFEMethod any more. If we do then we should promote them to protected static functions inside `FEMechanicsBase` so that we don't need to duplicate code.

boyceg replied:
These calls are needed to setup all of the vectors used in the different systems. I've renamed the key function `setup_system_vectors` and moved it into the base class as a static member function.

drwells commented on src/IB/IBFEMethod.cpp:
```diff
@@ -402,163 +417,71 @@ IBFEMethod::IBFEMethod(const std::string& object_name,
+    for (const std::string& vector_name : vector_names)
     {
-        d_PK1_stress_fcn_data[part].back().quad_type = d_default_quad_type[part];
+        convert_parallel_to_ghosted(vector_name, Phi_system);
```
Could we move this into the restart data load section somehow instead? IIRC this problem only occurs when we load vectors, not when we create them.

boyceg replied:
See above.

drwells commented on src/IB/IBFEMethod.cpp:
```diff
@@ -402,163 +417,71 @@ IBFEMethod::IBFEMethod(const std::string& object_name,
-    TBOX_ASSERT(part < d_num_parts);
-    d_lag_surface_force_fcn_data[part] = data;
+    // libMesh also stores a convenience pointer to the RHS that we need to reset:
+    Phi_system.rhs = &Phi_system.get_vector("RHS Vector");
```
same

boyceg replied:
See above.

drwells commented on src/IB/IBFEMethod.cpp:
```diff
@@ -402,163 +417,71 @@ IBFEMethod::IBFEMethod(const std::string& object_name,
+    const std::vector<std::string> vector_names{ "half", "tmp", "RHS Vector" };
+    for (const std::string& vector_name : vector_names)
+    {
+        convert_parallel_to_ghosted(vector_name, Q_system);
```
same

boyceg replied:
See above.

drwells commented on src/IB/IBFEMethod.cpp:
```diff
@@ -693,50 +568,36 @@ IBFEMethod::preprocessIntegrateData(double current_time, double new_time, int nu
-        // Update the coordinate mapping dX = X - s.
+    // Update the coordinate mapping dX = X - s.
+    for (unsigned part = 0; part < d_meshes.size(); ++part)
+    {
```
This doesn't seem right - we also call updateCoordinateMapping in `FEMechanicsBase::postprocessIntegrateData()`.

boyceg replied:
In the initial implementation, `IBFEMethod::pre/postprocessIntegrateData()` did not call the base class implementation.

If you prefer to leverage the base class implementation, that's OK with me, but then I think the base class should not do any data allocation --- for instance, because subclasses could be using different time stepping schemes and need different system vector instances. (This already happens even within `IBFEMethod` depending on whether one uses midpoint or trapezoidal rule, for instance.)

boyceg replied:
OK, so the base class now does essentially no data initialization or allocation. Once we have multiple implementations of the base class, perhaps we can think about whether it makes sense to factor some of that into the base class.

drwells commented on src/IB/IBFEMethod.cpp:
```diff
@@ -1393,37 +1258,8 @@ IBFEMethod::initializeFEEquationSystems()
-        const std::array<std::string, 2> system_names{ { COORDS_SYSTEM_NAME, VELOCITY_SYSTEM_NAME } };
-        const std::array<std::string, 2> vector_names{ { "new", "half" } };
+        const std::vector<std::string> system_names{ COORDS_SYSTEM_NAME, VELOCITY_SYSTEM_NAME };
+        const std::vector<std::string> vector_names{ "new", "half" };
```
Isn't all this stuff done in `FEMechanicsBase` now?

boyceg replied:
Nope; see above.

drwells commented on src/IB/IBFEMethod.cpp:
```diff
@@ -3638,24 +2872,32 @@ IBFEMethod::commonConstructor(const std::string& object_name,
+            d_fe_order_pressure[part] = SECOND;
             d_default_quad_order[part] = FIFTH;
+            d_default_quad_order_stress[part] = FIFTH;
+            d_default_quad_order_force[part] = FIFTH;
```
Shouldn't all of this also be in the base class now?

boyceg replied:
Dropped.

boyceg commented on include/ibamr/IBFEMethod.h:
```diff
@@ -1162,47 +862,45 @@ class IBFEMethod : public IBStrategy
+     * for the position.
      */
-    std::unique_ptr<IBTK::LibMeshSystemIBVectors> d_U_vecs;
+    IBTK::LibMeshSystemIBVectors* d_X_IB_vecs;
```
I just added extra `LibMeshSystemIBVectors` objects to the derived class. Seems easier than creating an interface --- maybe we can revisit after there are additional implementations.

boyceg commented:
@drwells, I think this now addresses most of your comments. I need to re-run the test suite and add a couple tests.

boyceg commented:
Looks like the usual MacOS suspects are failing:
```
The following tests FAILED:

    IBFE/explicit_ex4_2d.scratch_hier.mpirun=4.input
    IBFE/explicit_ex4_2d.postprocessor.mpirun=4.input
    IBFE/explicit_ex4_2d.scratch_hier.restart=25.mpirun=4.input
    IBFE/explicit_ex4_2d.inactive_1.mpirun=4.input
    IBFE/explicit_ex4_2d.inactive_1.mpirun=4.restart=50.input
    IBFE/explicit_ex4_2d.mpirun=4.input
    IBFE/explicit_ex1_2d.mpirun=4.input
    IBFE/explicit_ex4_2d.postprocessor.scratch_hier.mpirun=4.input
    IBFE/explicit_ex5_2d.regrid.mpirun=2.input
    IBFE/explicit_ex0_2d.input
```

boyceg commented:
Ah, `IBFE/explicit_ex1_2d.mpirun=4.input` and `IBFE/explicit_ex0_2d.input` should not have been failing for me --- found a bug; now fixed. Others it is hard to know, as they also fail for me with master.

drwells commented:
IBFE/explicit_ex4_2d.source_test.input fails for me but the rest of the tests currently pass.

boyceg commented:
> IBFE/explicit_ex4_2d.source_test.input fails for me but the rest of the tests currently pass.

Fails catastrophically or just gives different results?

drwells commented:
it fails at the diff stage.

boyceg commented:
Oh, it is because I named the input/output files incorrectly. Can you take another look?

drwells commented on include/ibamr/FEMechanicsBase.h:
```diff
@@ -0,0 +1,616 @@
+     * if (dump_restart_data && (iteration_num % restart_dump_interval == 0 || last_step))
+     * {
+     *     RestartManager::getManager()->writeRestartFile(restart_dump_dirname, iteration_num);
+     *     ib_method_ops->writeFEDataToRestartFile(restart_dump_dirname, iteration_num);
```
```suggestion
     *     fe_mechanics->writeFEDataToRestartFile(restart_dump_dirname, iteration_num);
```

boyceg replied:
👍

drwells commented on src/IB/IBFEMethod.cpp:
```diff
@@ -1278,1241 +1107,739 @@ void
+    // Set up the coupling matrix which will be used by each system.
+    d_diagonal_system_coupling.resize(NDIM);
+    for (unsigned int i = 0; i < NDIM; ++i)
+        for (unsigned int j = 0; j < NDIM; ++j) d_diagonal_system_coupling(i, j) = i == j ? 1 : 0;
```
`d_diagonal_system_coupling` is a member of the base class, so we should set it up inside a base class member function.

boyceg replied:
👍

drwells commented on src/IB/IBFEMethod.cpp:
```diff
@@ -1278,1241 +1107,739 @@ void
+    d_primary_fe_data_managers.resize(d_meshes.size(), nullptr);
+    d_scratch_fe_data_managers.resize(d_meshes.size(), nullptr);
+    d_active_fe_data_managers.resize(d_meshes.size(), nullptr);
+    d_fe_data.resize(d_meshes.size(), nullptr);
```
we don't need the `nullptr`s

boyceg replied:
👍 

I removed some others while I was at it...

drwells commented on src/IB/IBFEMethod.cpp:
```diff
@@ -1278,1241 +1107,739 @@ void
+    // IB ghost regions.
+    d_X_vecs.reset(new LibMeshSystemIBVectors(d_active_fe_data_managers, COORDS_SYSTEM_NAME));
+    d_U_vecs.reset(new LibMeshSystemIBVectors(d_active_fe_data_managers, VELOCITY_SYSTEM_NAME));
+    d_F_vecs.reset(new LibMeshSystemIBVectors(d_active_fe_data_managers, FORCE_SYSTEM_NAME));
```
We don't need to do this reset now.

drwells commented on src/IB/IBFEMethod.cpp:
```diff
@@ -2521,152 +1848,100 @@ IBFEMethod::assembleInteriorForceDensityRHS(PetscVector<double>& G_rhs_vec,
+                // N:     unit normal in the reference configuration
+                // n:     unit normal in the current configuration
+                // dA_da: reference surface area per current surface area (from Nanson's
+                //        relation)
```
:heavy_check_mark:

boyceg commented on include/ibamr/FEMechanicsBase.h:
```diff
@@ -0,0 +1,616 @@
+ * @brief      Class FEMechanicsBase provides core finite element mechanics
+ *             functionality and data management.
+ */
+class FEMechanicsBase : public SAMRAI::tbox::Serializable
```
You tell me. :-)

drwells replied:
I think the current version is fine. In particular, since restarts still work, what we have suffices.

boyceg commented on tests/IBFE/explicit_ex4_source_test.cpp:
```diff
@@ -0,0 +1,612 @@
+// ---------------------------------------------------------------------
```
Will do.

boyceg replied:
OK I added an option, but it also required some other changes to the example code --- not sure if that impacts outputs for other tests. Can you take another look?

boyceg commented on tests/IBFE/explicit_ex4_2d_source_test.output:
```diff
@@ -0,0 +1,300 @@
+  projecting the interpolated velocity field
+INSStaggeredHierarchyIntegrator::regridProjection(): regrid projection solve number of iterations = 0
+INSStaggeredHierarchyIntegrator::regridProjection(): regrid projection solve residual norm        = 0
+Total number of elems: 309
```
Interesting/horrifying.

boyceg replied:
This is setup to use `QUAD` elements; here's the logic to setup the mesh:
```
// NOTE: number of segments along boundary is 4*2^r.
            const double num_circum_segments = 2.0 * M_PI * R / ds;
            const int r = log2(0.25 * num_circum_segments);
            MeshTools::Generation::build_sphere(mesh, R, r, Utility::string_to_enum<ElemType>(elem_type));
```
FWIW, here is what is done for `TRI` elements:
```
            const int num_circum_nodes = ceil(2.0 * M_PI * R / ds);
            for (int k = 0; k < num_circum_nodes; ++k)
            {
                const double theta = 2.0 * M_PI * static_cast<double>(k) / static_cast<double>(num_circum_nodes);
                mesh.add_point(libMesh::Point(R * cos(theta), R * sin(theta)));
            }
            TriangleInterface triangle(mesh);
            triangle.triangulation_type() = TriangleInterface::GENERATE_CONVEX_HULL;
            triangle.elem_type() = Utility::string_to_enum<ElemType>(elem_type);
            triangle.desired_area() = 1.5 * sqrt(3.0) / 4.0 * ds * ds;
            triangle.insert_extra_points() = true;
            triangle.smooth_after_generating() = true;
            triangle.triangulate();
```
I think that most of the discrepancies are for `ex4` and `ex5`. I wonder if the root of the problem is that we are not getting consistent meshes?

drwells replied:
I think it might be due to the use of triangle - e.g., #876 had related problems that I could trace back to libMesh's patches to triangle.

boyceg commented:
I think I have addressed requested changes.

boyceg commented:
Should we remove Triangle from all tests?

drwells commented:
I think we already have at this point. However, I'm not completely convinced the failure is due to triangle - the change libMesh made only affected node numberings, not the actual mesh geometry.

boyceg commented:
It looks like `explicit_ex4.cpp` and `explicit_ex5.cpp` both use `TriangleInterface` to generate triangular meshes, and they both have input files that specify `TRI3` or `TRI6` elements.

drwells commented:
Hm. It must not be triangle, then.

drwells commented:
We should still get rid of it to get more consistent output for the same reasons specified in #876. Let me open an issue for that.

boyceg commented:
Aren't you getting different numbers of elements than me? Couldn't that be because ... Triangle is generating different meshes for us?

drwells commented:
Maybe. I am not completely convinced yet that Triangle is producing substantially different output since, IIRC, most ex4 tests do currently pass on macs. There must be a different explanation.

boyceg commented:
@drwells, if the only issue is the failing source test, how about you run it in Linux and update the output, and then we can worry about it later when trying to get the tests working on MacOS? I would like to get this merged so I can setup the FE mechanics solver.

drwells commented:
Sure.

drwells commented:
I force pushed a change to the test output file.

boyceg commented:
Command line git diff should be better.

drwells commented:
You're right - `diff -y` makes this a lot easier to read.

I think this is fine so lets merge.
